source/sessions/snapshot-sessions.rst

```diff
@@ -0,0 +1,358 @@
+    operations executed by an application that are related in some way. This
+    specification defines how sessions are used to implement snapshot reads.
+
+Snapshot reads
+    Reads with readconcern level ``snapshot`` that occur outside of transactions on
+    both the primary and secondary nodes, including in sharded clusters.
+
+Snapshot cluster time
+    Snapshot cluster time, representing timestamp of the first read (find/aggregate operations) in the session.
+    The server creates a cursor in response to a snapshot read command and 
+    reports ``atClusterTime`` field in the cursor response. ``atClusterTime`` field represents the timestamp
+    of the read and is guaranteed to be majority committed.
+
+Specification
+=============
+
+An application requests snapshot reads by creating a ``ClientSession``
+with options that specify that snapshot reads are desired. An
+application then passes the session as an argument to methods in the
+``MongoDatabase`` and ``MongoCollection`` classes. Read operations (find/aggregate/distinct) performed against
+that session will be read from the same snapshot.
+
+High level summary of the API changes for snapshot reads
+========================================================
+
+Snapshot reads are built on top of client sessions.
+
+Applications will start a new client session for snapshot reads like
+this:
+
+.. code:: typescript
+
+    options = new SessionOptions(isSnapshot = true);
+    session = client.startSession(options);
+
+All read operations performed using this session will be read from same snapshot.
+
+If no value is provided for ``isSnapshot`` a value of false is
+implied.
+
+MongoClient changes
+===================
+
+There are no API changes to ``MongoClient`` to support snapshot reads.
+Applications indicate whether they want snapshot reads by setting the
+``isSnapshot`` field in the options passed to the ``startSession`` method.
+
+SessionOptions changes
+======================
+
+``SessionOptions`` change summary
+
+.. code:: typescript
+
+    class SessionOptions {
+        Optional<bool> isSnapshot;
+
+        // other options defined by other specs
+    }
+
+In order to support snapshot reads a new property named
+``isSnapshot`` is added to ``SessionOptions``. Applications set
+``isSnapshot`` when starting a client session to indicate
+whether they want snapshot reads. All read operations performed
+using that client session will share the same snapshot.
+
+Each new member is documented below.
+
+isSnapshot
+---------
+
+Applications set ``isSnapshot`` when starting a session to
+indicate whether they want snapshot reads.
+
+Note that the ``isSnapshot`` property is optional. The default value of
+this property is false.
+
+Snapshot reads and causal consistency are mutually exclusive. Therefore if ``isSnapshot`` is set to true,
+``causalConsistency`` property is set to false. Client MUST throw an Error if both ``isSnapshot`` and ``causalConsistency`` are set to true.
```

Comment:
The server permits a snapshot read to **begin** with afterClusterTime so it's causally after some previous operation. Within a snapshot read, of course, you can't be causally consistent. This API doesn't expose the "begin a snapshot read causally after an operation" feature. That might be fine, and we could enhance the API later if it turns out that users want that feature. I want to confirm it's our intention to omit the feature now.

Reply:
Confirmed.

Reply:
@jyemin @ajdavis
Making sure that we want drivers forbid requesting both ``causalConsistency`` and ``isSnapshot`` (need to add a test for that).
I lean to raising an error in such case in driver, as otherwise due to drivers possible implementation differences it could be that some would send both `atClusterTime` and `afterClusterTime`, and some only one of the fields. Which will result in inconsistent behavior.

Reply:
Let's start with making it an error.  It's easier to go from error to not error than the other way around.

Reply:
Just FYI, libmongoc defines `mongoc_session_opt_t` with a field `flags` that is an enum [`mongoc_session_flag_t`](https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/src/mongoc/mongoc-client-session-private.h#L41-L44). I'll add another variant for `isSnapshot`, and by nature of the enum, there will be no way to specify `causalConsistency` and `isSnapshot` simultaneously. I'll still throw an error if users try to, but worth noting that if other drivers use a similar enum, users won't be able to do this anyways.